### PR TITLE
Extend RabbitMQ scaler to support count unacked messages

### PR DESCRIPTION
**Pull request by holyketzer**
Add settings which allows to count total messages count in queue (ready + unacked) to do not scale worker down if some long time processing is going.

By default `countUnacked = false` and it keeps old behaviour via AMQP protocol, if `countUnacked = true` it requires also `apiHost` to be set and uses HTTP management API to get queue info.

AMQP protocol doesn't allow to fetch unacked messages count, that is we need management API here. Technically we an completely switch to management API but as I understand not all servers have it enabled, that is why backward compatibility is saved and `countUnacked` is optional. 

### Checklist

- [x] A PR is opened to update the documentation on https://github.com/kedacore/keda-docs/pull/123
- [x] Tests have been added

It was tested in our Kubernetes cluster

Fixes https://github.com/kedacore/keda/issues/356

### CODE REVIEW

**Comment by tomkerkhove** [comment]
Docs are saying this:
> By default `countUnacked` is `false` in this case scaler uses AMQP protocol, requires `host` and only counts messages in the queue and ignores unacked messages.
If `countUnacked` is `true` then `host` is not required but `apiHost` is required in this case scaler uses HTTP management API and counts messages in the queue + unacked messages count.

Which requires following URLs:
- Host: `amqp://guest:password@localhost:5672/`
- Api Host: `https://guest:password@localhost:443/vhostname`

One could argue that we should always require `guest:password@localhost:5672` to be configured and change protocol based on the flags that are turned on?

Certainly since vhost is being added via https://github.com/kedacore/keda-docs/pull/70 (pending review) which is also the same.

File: pkg/scalers/rabbitmq_scaler.go [hunk 1]
```diff
@@ -26,9 +33,17 @@ type rabbitMQScaler struct {
...
 
 type rabbitMQMetadata struct {
-	queueName   string
-	host        string
-	queueLength int
+	queueName    string
+	host         string // connection string for AMQP protocol
+	apiHost      string // connection string for management API requests
+	queueLength  int
+	countUnacked bool // if true uses HTTP API and requires apiHost, if false uses AMQP and requires host
```
Comment (tomkerkhove) [inline pkg/scalers/rabbitmq_scaler.go]: What about `includeUnacked`? Now it feels like it will only include unacked messages

**Comment by holyketzer** [comment]
> One could argue that we should always require guest:password@localhost:5672 to be configured and change protocol based on the flags that are turned on?

Sure I tried it first and suddenly discovered that our credentials for AMQP protocol has different permissions level and can't be used to access management API. That is why I introduced new connection string

`Certainly since vhost is being added via kedacore/keda-docs#70 (pending review) which is also the same.`

Hm, we have been using vhost in AMQP connection string since we started to use Keda and it works )

File: pkg/scalers/rabbitmq_scaler.go (hunk 1 above)
Comment (holyketzer) [reply pkg/scalers/rabbitmq_scaler.go]: that makes sense

File: pkg/scalers/rabbitmq_scaler.go [hunk 2]
```diff
@@ -105,10 +149,12 @@ func getConnectionAndChannel(host string) (*amqp.Connection, *amqp.Channel, erro
 
 // Close disposes of RabbitMQ connections
 func (s *rabbitMQScaler) Close() error {
-	err := s.connection.Close()
-	if err != nil {
-		rabbitmqLog.Error(err, "Error closing rabbitmq connection")
-		return err
+	if s.metadata.countUnacked == false {
```
Comment (ahmelsayed) [inline pkg/scalers/rabbitmq_scaler.go]: how about `s.connection != nil` so it'd make sure to dispose the connection if it's at all created.

**Comment by ahmelsayed** [comment]
I'm not very familiar with rabbitmq, so not sure I have anything to add to the discussion. I think having a single `host` as opposed to `apiHost` and `host` but from the description it doesn't seem that they are always compatible? could they be different? can the same user account work with both given the right permissions?

**Comment by tomkerkhove** [comment]
> > One could argue that we should always require guest:password@localhost:5672 to be configured and change protocol based on the flags that are turned on?
> 
> Sure I tried it first (parse AMQP connection string, replace protocol, port, extract vhost) and suddenly discovered that our credentials for AMQP protocol has different permissions level and can't be used to access management API. That is why I introduced new connection string.
> 
> We can use all different fields for like `host`, `port`, `user`, `password` and `vhost` and share them, or maybe yes, we can use one `host` setting in different contexts as AMQP or HTTP protocol, but i'm not sure what is more clear, maybe one `host` is better than `host` and `apiHost`

I think using `host` would be better but that would be a breaking change, so let's stick with seperation between the 2 and we'll open an issue to change this for 2.0 - What do you think @zroubalik @ahmelsayed ?

> > Certainly since vhost is being added via [kedacore/keda-docs#70](https://github.com/kedacore/keda-docs/pull/70) (pending review) which is also the same.
> 
> Hm, we have been using vhost in AMQP connection string since we started to use Keda and it works ). You mean it wasn't documented?

Yes, I've merged it now

**Comment by zroubalik** [comment]
> I'm not very familiar with rabbitmq, so not sure I have anything to add to the discussion. I think having a single `host` as opposed to `apiHost` and `host` but from the description it doesn't seem that they are always compatible? could they be different? can the same user account work with both given the right permissions?

+1 good question

I agree with @tomkerkhove point on opening a new issue targeted for 2.0 for the breaking change (in case it makes sense to implement that).

**Comment by holyketzer** [comment]
Ok, I got all the points and going to fix all of them.

**Comment by holyketzer** [comment]
Feedback is accepted I pushed fixes + fix for doc PR

**Comment by zroubalik** [comment]
@ahmelsayed should we merge this one? WDYT?

**Comment by ahmelsayed** [comment]
Thanks @holyketzer for your contributions!

**Comment by holyketzer** [comment]
Cool thank you guys!

**Pull request merged**